exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,521 @@
+
+* A single observation function, `sols: 'a m -> 'a Seq.t`, which
+  converts a computation to a sequence of results, thereby allowing
+  the user to execute this computation and observe its results. The
+  name `sols` stands for `solutions`.
+
+A monad can be thought of as a **mini-programming language** where
+computations are first-class citizens: we have a type of
+computations, ways of building computations, and a way of executing
+computations.
+
+A computation in the nondeterminism monad can produce zero, one, or
+more results. Indeed, a computation that fails produces zero
+results. A computation that succeeds normally produces one result.
+A computation that uses `choose` can produce more than one result.
+It is in fact possible to construct computations that produce an
+infinite number of results!
+
+Thus, a useful way to think of a computation is as **a sequence of
+results**, that is, a sequence of results.
+
+Because of this remark, one might be tempted to define the type `'a
+m` as a synonym for `'a Seq.t`, the type of sequences of values of
+type `'a`. However, although it is possible to represent a
+computation internally as a sequence (and we will do so in this
+exercise), this is not necessarily the best implementation
+technique. Thus, we prefer to make the API more flexible by viewing
+`'a m` as an abstract type and by offering an observation function,
+`sols`, which converts a computation to a sequence.
+
+## The Nondeterminism Monad's API
+
+The signature, or API, of the nondeterminism monad is as follows:
+
+```
+  (* Type. *)
+  type 'a m
+
+  (* Constructor functions. *)
+  val return: 'a -> 'a m
+  val (>>=): 'a m -> ('a -> 'b m) -> 'b m
+  val fail: 'a m
+  val choose: 'a m -> 'a m -> 'a m
+  val delay: (unit -> 'a m) -> 'a m
+  val at_most_once: 'a m -> 'a m
+  val interleave: 'a m -> 'a m -> 'a m
+  val (>>-): 'a m -> ('a -> 'b m) -> 'b m
+
+  (* Observation function. *)
+  val sols: 'a m -> 'a Seq.t
+```
+
+As explained above, a value of type `'a m` is **a description of a
+computation**, which, once executed, produces a sequence of results
+of type `'a`.
+
+To execute a computation `m`, one must first convert it to a
+sequence of type `'a Seq.t`, whose elements can then be demanded,
+one by one. (More information on the module `Seq` is given below.)
+This conversion is performed by the observation function `sols`.
+
+The call `sols m` typically terminates in constant time; the actual
+computation described by `m` takes place only when the elements of
+the sequence `sols m` are demanded, and only insofar as necessary to
+produce the elements that are demanded. For instance, applying
+`Seq.head` to the sequence `sols m` forces the computation to
+proceed up to the point where it is able to produce its first
+result.
+
+The constructor functions `return` and `(>>=)` exist in all monads.
+(They are also known as `return` and `bind`.) `return` constructs a
+trivial computation, which does nothing except return a value,
```

Comment:
I would say "except returning", but you usually have far better English than mine :).

Reply:
Thanks, changed to "nothing but", which sounds more idiomatic to me.